Spigot-Server-Patches/0418-Ticket-level-per-player-view-distance-API.patch

```diff
@@ -0,0 +1,186 @@
++                this.a(j, b0, this.c(b0), b0 <= i - 2, i); //Paper ticket level pp-view-dist
+             }
+
+             this.e = i;
+         }
+
+-        private void a(long i, int j, boolean flag, boolean flag1) {
+-            if (flag != flag1) {
++        private void a(long i, int j, boolean flag, boolean flag1, int vd) { //Paper ticket level pp-view-dist
++            if (flag != flag1 || (flag && flag1)) { //Paper ticket level pp-view-dist
+                 Ticket<?> ticket = new Ticket<>(TicketType.PLAYER, ChunkMapDistance.b, new ChunkCoordIntPair(i), ChunkMapDistance.this.currentTick);
+
+                 if (flag1) {
+                     ChunkMapDistance.this.j.a(ChunkTaskQueueSorter.a(() -> { // Craftbukkit - decompile error
+                         ChunkMapDistance.this.m.execute(() -> {
+-                            if (this.c(this.c(i))) {
++                            if (this.getChunksMapping(i) <= vd - 2) { //Paper ticket level pp-view-dist
+                                 ChunkMapDistance.this.addTicket(i, ticket);
+                                 ChunkMapDistance.this.l.add(i);
+                             } else {
+@@ -412,10 +414,14 @@ public abstract class ChunkMapDistance {
+         public void a() {
+             super.a();
+             if (!this.g.isEmpty()) {
+-                LongIterator longiterator = this.g.iterator();
++                //Paper start ticket level pp-view-dist
++                ObjectIterator<Long2ByteMap.Entry> longiterator = this.getUpdateChunks().long2ByteEntrySet().iterator();
+
+                 while (longiterator.hasNext()) {
+-                    long i = longiterator.nextLong();
++                    Long2ByteMap.Entry e = longiterator.next();
++                    long i = e.getLongKey();
++                    int vd = e.getByteValue();
++                    //Paper end
+                     int j = this.f.get(i);
+                     int k = this.c(i);
+
+@@ -430,7 +436,7 @@ public abstract class ChunkMapDistance {
+                             }
+
+                         });
+-                        this.a(i, k, this.c(j), this.c(k));
++                        this.a(i, k, this.c(j), k <= vd - 2, vd); //Paper ticket level pp-view-dist
+                     }
+                 }
+
+@@ -448,6 +454,8 @@ public abstract class ChunkMapDistance {
+
+         protected final Long2ByteMap a = new Long2ByteOpenHashMap();
+         protected final int b;
++        protected int pvd; //Paper ticket level pp-view-dist
++        protected int getChunksMapping(long i) {return this.c(i);} //Paper OBFHELPER
+
+         protected b(int i) {
+             super(i + 2, 16, 256);
+@@ -464,16 +472,31 @@ public abstract class ChunkMapDistance {
+         protected void a(long i, int j) {
+             byte b0;
+
++            //Paper start ticket level pp-view-dist
++            if(j == 0) {
++                ObjectSet<EntityPlayer> c = ChunkMapDistance.this.getChunkPlayers(i);
```

Comment:
this maps to retrieving a set of players at the specific chunk coordinate, not players in range

meaning this code is broken

Reply:
no? that is the point of c and that is how it is used here exactly? to check for whether the player is in the chunk? this method doesn't check for anything in range, it utilizes the full potential of the input stream of the setLevel function itself

Reply:
the function this is invoked in is the function noting a change in distance for a specific chunk. the change is then queued to be processed via the a() method which you have correctly modified.
however you're only getting the view distance of players in that chunk, so this breaks for chunks which should be in range of players but are not occupied by players

this is the most important part, which is why I said mojang's system does not support it: it does not map players to chunks they are in range of

Reply:
I remember accounting for this before, and I think this method avoids this. I suggest you load and debug the input stream.

Reply:
possible to clarify what you mean by 'the function this is invoked in is the function noting a change in distance...'? as it called to set/refresh the entire map during a full refresh/update of all chunk levels, and this full refresh is called whenever a player moves/enters a new chunk, instead of a singular change.

Reply:
this function is invoked when a chunk's distance changes from a player, in a radius of 33

Reply:
correct, but the invoking function re-iterates through every single chunk level in the queue again through this function, instead of a single change. this is what this method takes advantage of, and attaches view distance per player according to how the queue is structured.

Reply:
I will need to look into this

Reply:
One issue I can see is that this might not work for overlapping areas, since your method would set the view distance to an unexpected value one you hit a chunk with another player. 

I would also prefer to not rely on specific implementation detail when we have alternatives already technically in paper for tracking players and view distance.

Reply:
This has been accounted for and solved through some rather simple mathematics. A commit will be pushed soon.